File: solr/core/src/java/org/apache/solr/util/pf4j/Plugins.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.solr.util.pf4j;

import java.lang.invoke.MethodHandles;
import java.util.Collections;
import java.util.List;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import com.github.zafarkhaja.semver.Version;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import ro.fortsoft.pf4j.PluginManager;
import ro.fortsoft.pf4j.update.UpdateManager;
import ro.fortsoft.pf4j.update.UpdateRepository;

import static ro.fortsoft.pf4j.update.UpdateRepository.PluginInfo;
import static ro.fortsoft.pf4j.update.UpdateRepository.PluginRelease;

/**
 * Discovers and loads plugins from plugin folder using PF4J
 */
public class Plugins {
  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());

  private final SolrPluginManager pluginManager;
  private final PluginUpdateManager updateManager;
  private final Version systemVersion;


  public Plugins() {
    pluginManager = new SolrPluginManager();
    updateManager = new PluginUpdateManager(pluginManager);
    systemVersion = pluginManager.getSystemVersion();
  }

  public void load() {
    pluginManager.loadPlugins();
    pluginManager.startPlugins();
  }

  private static Predicate<UpdateRepository.PluginInfo> filterPredicate(String q) {
    return p -> (q == null || q.length() == 0 || q.equals("*")) || p.id != null && p.id.contains(q);
  }

  public List<UpdateRepository.PluginInfo> query(String q) {
    // TODO: Allow install from any GitHub repo
    if (updateManager.hasAvailablePlugins()) {
      List<UpdateRepository.PluginInfo> plugins = updateManager.getAvailablePlugins()
          .stream().filter(filterPredicate(q)).collect(Collectors.toList());
      log.debug("Found these plugins in repos for query " + q + ": " + plugins);
      log.info("Found plugins for " + q + ": " + plugins.stream().map(p -> p.id +
          "(" + p.getLastRelease(systemVersion).version + ")").collect(Collectors.toList()));
      return plugins;
    } else {
      return Collections.emptyList();
    }
  }

  public List<String> listInstalled() {
    List<String> plugins = pluginManager.getPlugins().stream().map(p -> p.getDescriptor().toString()).collect(Collectors.toList());
    log.info("Currently installed plugins: " + plugins);
    return plugins;
  }

  public void updateAll() {
    if (updateManager.hasUpdates()) {
      List<UpdateRepository.PluginInfo> updates = updateManager.getUpdates();
      for (PluginInfo plugin : updates) {
        PluginRelease lastRelease = plugin.getLastRelease(systemVersion);
        String lastVersion = lastRelease.version;
        String installedVersion = pluginManager.getPlugin(plugin.id).getDescriptor().getVersion().toString();
        // TODO: Inspect whether we can use the plugin or not
        log.info("Updating plugin with id " + plugin.id);
        updateManager.updatePlugin(plugin.id, lastRelease.url);
    }
    }
  }

  public void install(String id) {
    log.info("Installing plugin with id "+id);
    PluginInfo info = query(id).get(0);
    updateManager.installPlugin(info.getLastRelease(systemVersion).url);
  }

  public PluginManager getPluginManager() {
    return pluginManager;
  }

  public ClassLoader getUberClassLoader(ClassLoader parent) {
    return pluginManager.getUberClassloader(parent);
  }
  
  public void addUpdateRepository(String id, String url) {
    updateManager.addRepository(id, url);
  }

  public UpdateManager getUpdateManager() {
    return updateManager;
  }

  public void uninstall(String id) {
    updateManager.uninstallPlugin(id);
  }
}

File: solr/webapp/angularJS/js/angular/controllers/plugins.js
/*
 Licensed to the Apache Software Foundation (ASF) under one or more
 contributor license agreements.  See the NOTICE file distributed with
 this work for additional information regarding copyright ownership.
 The ASF licenses this file to You under the Apache License, Version 2.0
 (the "License"); you may not use this file except in compliance with
 the License.  You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
*/

solrAdminApp.controller('PluginsController',
    function($scope, $rootScope, $routeParams, $location, Mbeans, Constants) {
        $scope.resetMenu("plugins", Constants.IS_CORE_PAGE);

        if ($routeParams.legacytype) {
            // support legacy URLs. Angular cannot change #path without reloading controller
            $location.path("/"+$routeParams.core+"/plugins");
            $location.search("type", $routeParams.legacytype);
            return;
        }

        $scope.refresh = function() {
            Mbeans.stats({core: $routeParams.core}, function (data) {
                var type = $location.search().type;
                $scope.types = getPluginTypes(data, type);
                $scope.type = getSelectedType($scope.types, type);

                if ($scope.type && $routeParams.entry) {
                    $scope.plugins = $routeParams.entry.split(",");
                    openPlugins($scope.type, $scope.plugins);
                } else {
                    $scope.plugins = [];
                }
            });
        };

        $scope.selectPluginType = function(type) {
            $location.search({entry:null, type: type.lower});
            $scope.type = type;
        };

        $scope.selectPlugin = function(plugin) {
            plugin.open = !plugin.open;

            if (plugin.open) {
                $scope.plugins.push(plugin.name);
            } else {
                $scope.plugins.splice($scope.plugins.indexOf(plugin.name), 1);
            }

            if ($scope.plugins.length==0) {
                $location.search("entry", null);
            } else {
                $location.search("entry", $scope.plugins.join(','));
            }
        }

        $scope.startRecording = function() {
            $scope.isRecording = true;
            Mbeans.reference({core: $routeParams.core}, function(data) {
                $scope.reference = data.reference;
                console.log($scope.reference);
            })
        }

        $scope.stopRecording = function() {
            $scope.isRecording = false;
            console.log($scope.reference);
            Mbeans.delta({core: $routeParams.core}, $scope.reference, function(data) {
                parseDelta($scope.types, data);
            });
        }

        $scope.refresh();
    });

var getPluginTypes = function(data, selected) {
    var keys = [];
    var mbeans = data["solr-mbeans"];
    for (var i=0; i<mbeans.length; i+=2) {
        var key = mbeans[i];
        var lower = key.toLowerCase();
        var plugins = getPlugins(mbeans[i+1]);
        if (plugins.length == 0) continue;
        keys.push({name: key,
                   selected: lower == selected,
                   changes: 0,
                   lower: lower,
                   plugins: plugins
        });
    }
    keys.sort(function(a,b) {return a.name > b.name});
    return keys;
};

var getPlugins = function(data) {
    var plugins = [];
    for (var key in data) {
        var pluginProperties = data[key];
        var stats = pluginProperties.stats;
        delete pluginProperties.stats;
        for (var stat in stats) {
            // add breaking space after a bracket or @ to handle wrap long lines:
            stats[stat] = new String(stats[stat]).replace( /([\(@])/g, '$1&#8203;');
        }
        plugin = {name: key, changed: false, stats: stats, open:false};
        plugin.properties = pluginProperties;
        plugins.push(plugin);
    }
    plugins.sort(function(a,b) {return a.name > b.name});
    return plugins;
};

var getSelectedType = function(types, selected) {
    if (selected) {
        for (var i in types) {
            if (types[i].lower == selected) {
                return types[i];
            }
        }
    }
};

var parseDelta = function(types, data) {

    var getByName = function(list, name) {
        for (var i in list) {
            if (list[i].name == name) return list[i];
        }
    }

    var mbeans = data["solr-mbeans"]
    for (var i=0; i<mbeans.length; i+=2) {
        var typeName = mbeans[i];
        var type = getByName(types, typeName);
        var plugins = mbeans[i+1];
        for (var key in plugins) {
            var changedPlugin = plugins[key];
            if (changedPlugin._changed_) {
                var plugin = getByName(type.plugins, key);
                var stats = changedPlugin.stats;
                delete changedPlugin.stats;
                plugin.properties = changedPlugin;
                for (var stat in stats) {
                    // add breaking space after a bracket or @ to handle wrap long lines:
                    plugin.stats[stat] = new String(stats[stat]).replace( /([\(@])/g, '$1&#8203;');
                }
                plugin.changed = true;
                type.changes++;
            }
        }
    }
};

var openPlugins = function(type, selected) {
    for (var i in type.plugins) {
        var plugin = type.plugins[i];
        plugin.open = selected.indexOf(plugin.name)>=0;
    }
}
